types/react/index.d.ts

```diff
@@ -90,6 +90,32 @@ declare namespace React {
     type RefCallback<T> = { bivarianceHack(instance: T | null): void }["bivarianceHack"];
     type Ref<T> = RefCallback<T> | RefObject<T> | null;
     type LegacyRef<T> = string | Ref<T>;
+    /**
+     * Gets the instance type for a React element. The instance will be different for various component types:
+     *
+     * - React class components will be the class instance. So if you had `class Foo extends React.Component<{}> {}`
+     *   and used `React.ElementRef<typeof Foo>` then the type would be the instance of `Foo`.
+     * - React stateless functional components do not have a backing instance and so `React.ElementRef<typeof Bar>`
+     *   (when `Bar` is `function Bar() {}`) will give you the `undefined` type.
+     * - JSX intrinsics like `div` will give you their DOM instance. For `React.ElementRef<'div'>` that would be
+     *   `HTMLDivElement`. For `React.ElementRef<'input'>` that would be `HTMLInputElement`.
+     *
+     * `C` must be the type _of_ a React component so you need to use typeof as in React.ElementRef<typeof MyComponent>.
+     */
+    type ElementRef<
+        C extends
+            | ComponentClass<any>
+            | FunctionComponent<any>
+            | keyof JSX.IntrinsicElements
+    > = C extends ComponentClass<any>
+        ? InstanceType<C>
+        : C extends FunctionComponent<any>
+        ? undefined
+        : C extends keyof JSX.IntrinsicElements
+        ? JSX.IntrinsicElements[C] extends DOMAttributes<infer E>
```

Comment:
This may not work for SVGs, I would instead try to retrieve the `Ref` itself:
```ts
type ElementRef<C> = C extends JSX.IntrinsicElements ? JSX.IntrinsicElements[C] extends { ref?: Ref<infer T> } ? T : never : never
```

Reply:
It works, see [the below test](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/43201/files#diff-f000af0f82be30333d108a5acb499a2aR446) or [this playground](https://www.typescriptlang.org/play/index.html?jsx=1#code/JYWwDg9gTgLgBAJQKYEMDG8BmUIjgIilQ3wChSYBPMJOAUQBskQkA7GZTAHlLj7gDCcJAA8YbACYBnXvzkAfRMRgA6Abkis2MAQxRSpXFK0oA+WXL6Lk6VQDEArqwzAIrdeDfajJ85f6KANZIlBCYcABSAMoAGioAkuxQwKxSwGiMzNoypnAAvILCYpJSSrZqGl7suvqGxmYWcAD8cIlSMMZoSAAq1EhcAn78AFyFouKs0mUYKo7OMK7ulVrsPg1yLU4SSJgpSBKNo0LjJXDBoeHRcYkwyanpmSzsMhuRsQlJKWkZTE8wUgBtAQAXSKEymNhmABEAPIAWQAgjBbsAAEYOcSGFKYJBQehDfwtOiNOSjLQAN1xhzgFNxAG5yGg9AYlJgPJptGDTpDVOyqvAAN4AX3ImCcLjcrIEKCgUEoKQA5nyVjAABRgHBgKSjeoASjgAsaRBgDigrDgXAkwHJpi4AHorTbSCKKH0pcttAipJx8vRftpOFwqDQwu7PCrTHS4Ha7XAACR0EQ0DC9GhhjnsV1pzjS2Xy1hKj3sL0+gqPAM7IN9UM5mVyxXK7SR6OxhNJpApt1bHZ7A7B2gACW6cIAMjc7t8SztfeX2IGAOSO+fNmPxxPJmCpwfDkdQ62zmBZ2hRABqAHFx190lPwmX-XPK-OpOSFcuo6u2xut3BT2ffwegA).